src/main/java/com/minecolonies/blockout/views/DropDownList.java

```diff
@@ -0,0 +1,308 @@
+        {
+            setSelectedIndex(0);
+        }
+        else
+        {
+            setSelectedIndex((selectedIndex + 1) % dataProvider.getElementCount());
+        }
+    }
+
+    /**
+     * Set the data provider to fill the list.
+     *
+     * @param p is the data provider for the list.
+     */
+    public void setDataProvider(final DataProvider p)
+    {
+        dataProvider = p;
+        list.setDataProvider(new ScrollingList.DataProvider()
+        {
+            @Override
+            public int getElementCount()
+            {
+                return dataProvider.getElementCount();
+            }
+
+            @Override
+            public void updateElement(final int index, @NotNull final Pane rowPane)
+            {
+                updateDropDownItem(rowPane, index, dataProvider.getLabel(index));
+            }
+        });
+
+        refreshElementPanes();
+    }
+
+    /**
+     * Use the data provider to update all the element panes.
+     */
+    public void refreshElementPanes()
+    {
+        list.refreshElementPanes();
+        if (list.getContentHeight() < dropDownHeight)
+        {
+            list.setSize(dropDownWidth, list.getContentHeight());
+        }
+        else
+        {
+            list.setSize(dropDownWidth, dropDownHeight);
+        }
+    }
+
+    /**
+     * open the dropdown list.
+     */
+    public void open()
+    {
+        refreshElementPanes();
+        overlay.setVisible(true);
+        overlay.setFocus();
+    }
+
+    /**
+     * close the dropdown list.
+     */
+    public void close()
+    {
+        overlay.setVisible(false);
+    }
+
+    /**
+     * Interface for a data provider that updates pane scrolling list pane info.
+     */
+    public interface DataProvider
+    {
+        int getElementCount();
+
+        String getLabel(final int index);
+    }
+
+    /**
+     * Update an pane item in the list.
+     *
+     * @param rowPane which need the update
+     * @param index of the item
+     * @param label use for this item
+     */
+    private void updateDropDownItem(@NotNull final Pane rowPane, final int index, final String label)
+    {
+        final Button choiceButton = rowPane.findPaneOfTypeByID("button", Button.class);
+        if (choiceButton != null)
+        {
+            // is idLabel necessary ?
+            final Label idLabel = rowPane.findPaneOfTypeByID("id", Label.class);
+            if (idLabel != null)
+            {
+                idLabel.setLabelText(Integer.toString(index));
+            }
+            choiceButton.setLabel(label);
+            choiceButton.setHandler(this);
+       }
+    }
+
+    /**
+     * Set the button handler for this button.
+     *
+     * @param h The new handler.
+     */
+    public void setHandler(final Handler h)
+    {
+        handler = h;
+    }
+
+    /**
+     * Used for windows that have dialog message.
+     */
+    @FunctionalInterface
+    public interface Handler
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Drop this interface in favor of "java.util.function.ObjIntConsumer<DropDownList>". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1711)